### ISSUE TITLE
Core: Recovery files ignored if project file modification date is newer, even if crash corrupted the project file

### ISSUE THREAD

**Issue by axxel** [FreeCAD/FreeCAD#18044]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

After hours of work on a pretty simple model with the 1.0 release on Linux, FC crashed with a SIGSEGV. That happened before and I thought I might have lost a couple of minutes since my last save. Restarting FC gave the following on the console:
```
FreeCAD 1.0.0, Libs: 1.0.0R39109 (Git)
(C) 2001-2024 FreeCAD contributors
FreeCAD is free and open-source software licensed under the terms of LGPL2+ license.

Ignore recovery file  "/home/axel/.cache/FreeCAD/Cache/FreeCAD_Doc_5b799e8d-5c96-4109-9b9d-7b1098506e98_9ed579_617729/fc_recovery_file.fcstd"  because it is older than the project file "/home/axel/Downloads/DieFr\xC3\xA4se.FCStd" 

Fatal Error at file /home/axel/Downloads/DieFräse.FCStd, line 155, char 27
Invalid Document.xml: unexpected end of input
Fatal Error at file GuiDocument.xml, line 1, char 1
Exception message is: 
invalid document structure
Reading failed from embedded file: GuiDocument.xml (0 bytes, 2 bytes compressed)
```
So I ended up with an empty project and by the time I noticed what happened, the `*.FCBak` file as well as the recovery file from the cache have been overridden with the empty file as well. So all my work was gone. This is **bad**.

Independent of the crash itself, which might have happened when I pressed save or something else, I don't remember, I would think that dismissing the recovery from a backup file only based on the timestamp is not good and the decision has to be deferred until after the file has been successfully loaded. And if that did not work, then doing a recovery is the only sensible approach.

To reproduce this issues, simply open any project, do some work until a recovery file is written to the cache. Then kill the FC process and manually corrupt the project file, e.g. truncate it. Start FC again, open the corrupted file and see what happens.

### Full version info

```shell
OS: Ubuntu 24.04.1 LTS (ubuntu:GNOME/ubuntu/xcb)
Architecture: x86_64
Version: 1.0.0.39109 (Git) Conda AppImage
Build type: Release
Branch: (HEAD detached at 1.0.0)
Hash: 2fcc5317fe3aee96ca73475986a577719fc78e20
Python 3.11.9, Qt 5.15.13, Coin 4.0.3, Vtk 9.2.6, OCC 7.7.2
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Dark.qss/FreeCAD Dark/Fusion
Installed mods: 
  * Defeaturing 1.2.2
  * A2plus 0.4.68
```

### Subproject(s) affected?

None

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by luzpaz** [comment]
Can anyone reproduce ?

**Comment by tarman3** [comment]
- Do some work until a recovery file is written to the cache.
- Then kill the FC process
- Manually corrupt the project file, truncate it.
- Start FC again and open project.

```
Fatal Error at file /home/user/Unnamed.FCStd, line 1, char 1
Ignore recovery file  "/home/user/.cache/FreeCAD/Cache/

FreeCAD_Doc_f9772b58-2d78-4ab9-8fa9-30faf70a9e37_ddad8b_305756/fc_recovery_file.fcstd"  because it is older than the project file "/home/user/Unnamed.FCStd"
```

**Comment by axxel** [comment]
> Also, experience with FreeCAD has developed a reflex to press CTRL+S almost before every action ))

I'm new to FreeCAD but can confirm that my fingers already start doing that by themselfs :-/.

And thanks for the tip about changing the number of FCBak files. I didn't know about that option. If that would have been at least 2, I assume my work would not have been lost completely.

**Comment by leoheck** [comment]
> reflex to press CTRL+S almost before every action ))
 
[Off topic] You should do this for every project on every software to be safe. My PhD advisor used to write a thesis before hitting the simple Ctrl+s shortcut. Who else is is stupid like that?

**Comment by hyarion** [comment]
This happened for a user on discord and has shared the file for examination ([asfdrsfg.FCStd.zip](https://github.com/user-attachments/files/18034295/asfdrsfg.FCStd.zip) <-- note .zip added to be able to upload and examine the file)

All the files in the archive is zero bytes except the xml which is cut short where a name should begin in a `ObjectDeps` tag.

---

What if we did the following when saving files:
1. Copy old file to a temporary file with `.old` suffix (or similar)
2. Overwrite the new file
3. Verify the new file that it is ok (in this case xml should be parseable and some checks against zero bytes files)
4. If it is verified to be ok, remove the old file
5. If not, show an error to the user that the file couldn't be saved and rename the old-file to original name (maybe add a way to keep the probably-broken file in case of a false negative verification)

**Comment by axxel** [comment]
> What if we did the following when saving files: [...]

If you want to tackle the issue while saving (instead of during recovery), it would be easier if FC wrote the new file with a temporary filename first, leaving the old one alone. Then after the write is complete, replace the old file with the new one. There would be no need to check the validity of the new file's content to deal with my reported problem, as in this case having written it without crashing is already proof enough that it worked.

I assume this approach is easier to implement than changing the recovery logic to first load the file and then decided whether or not a subsequent recovery is appropriate.

On a side note: the crash itself happened again. It was definitively during saving and I believe it had something to do with adding a link to a body with a non-zero element count and then undoing this. I would have to further investigate.

**Comment by bieli** [comment]
Unfortunately, I had yesterday similar issue - completelly dataloss from my project (one month work!).
I share with you my files with data loss. Strange thing is, that main project file and backup have inside (this is compressed) **0 sized files**!!! 

In my opinion strategy, how to save buckup - in 1.0.0 release - files has BUG.
In previous versions like 0.22-dev and 0.21, I seen files recovery dialog, but today nothing when FreeCAD start.

Terrible story is, that I follow by mainstream and use 1.0.0 as production - I know, this is Open Source.
My intention with below sentences is support community with data attached to this post, instead of negative feedback.

If it will be possible to recovery files, it saved me one month of work on project.
Maybe someone from FreeCAD' community would like to try!

[overritten_dataloss_in_FreeCAD_files_after_crash.tar.gz](https://github.com/user-attachments/files/18040984/overritten_dataloss_in_FreeCAD_files_after_crash.tar.gz)

### Some details about machine and OS
#### This is my release:
FreeCAD 1.0.0, Libs: 1.0.0R38641 +468 (Git)
(C) 2001-2024 FreeCAD contributors

#### OS:
Linux Ubuntu 20.04.6 LTS

#### Linux kernel:
Linux machine-name 5.15.84-051584-generic #202212190739 SMP Mon Dec 19 12:59:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

#### CPUinfo:
Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

#### RAM:
32GB

**Comment by hyarion** [comment]
It dies when trying to print name for @bieli too

**Comment by bieli** [comment]
@hyarion It was random die. No special thing, maybe I using Draw workspace with Part and switching between (it was my last action, as I remember). Next it's crashed without special stderr/stdout and without crash report (or I can't find it).

BUT I found something maybe **important** for people with the same story:
I found in properties in FreeCAD - thanks for engineers - in `Edit -> Preferences -> Cache` location, go to cache and I can see my project from last recovery or iteration. Yeeeaaah, I got it - this is my file in subfolder inside `/home/<USERNAME>/snap/freecad/common/temp/` with pattern `FreeCAD_Doc_ba7312e2-aaf5-4536-b003-0f514a668d09_45081c_540968/fc_recovery_file.fcstd`.

When I try to open this file, I can see my previous iteration of project !!! 
Thanks Open Source Engineers :-)

**Comment by axxel** [comment]
> Yeeeaaah, I got it

Good for you! When I discovered this file, it has already been overwritten with the then empty project. But I only lost 2-3 hours worth of work. I can highly recommend to increase the setting `Edit -> Preferences -> General -> Document -> Maximum number of backup files to keep when resaving document` to at least 2 as @tarman3 suggested above.

**Comment by bieli** [comment]
I'll put a PR with some idea on top of this subject to FreeCAD for next end-user, who will have similar situations!

**Comment by bieli** [comment]
Ok, I know more about this terrible error - this still exists, when I add more BAK files count then 1 (currently I put 3). 
When we setup more backups counts, it's not a solution, because backups are impacted/broken, too. 
This is exception log from my todays again crash with zeros in my project file. Maybe someone find bug here?
```
bash
$ freecad
FreeCAD 1.0.0, Libs: 1.0.0R38641 +468 (Git)
(C) 2001-2024 FreeCAD contributors
FreeCAD is free and open-source software licensed under the terms of LGPL2+ license.

Adding snap-specific PYTHONPATH to sys.path: /home/bieli/snap/freecad/common/.local/lib/python3.10/site-packages:/snap/freecad/1202/lib/python3.10/site-packages:/snap/freecad/1202/usr/lib/python3/dist-packages
Qt: Session management error: None of the authentication protocols specified are supported
3DfindIT: FreeCAD below version 0.19 is not supported. Please update to a recent version.
connect failed: No such file or directory
Fillet012: BRep_API: command not done
Fillet008: BRep_API: command not done
Fillet012: BRep_API: command not done
Fillet008: BRep_API: command not done
Fillet012: BRep_API: command not done
Fillet008: BRep_API: command not done
Fillet012: BRep_API: command not done
Fillet008: BRep_API: command not done
<Exception> invalid container
<App> Transactions.cpp(186): Exception on undo 'Paste':invalid container
Fillet012: BRep_API: command not done
Fillet008: BRep_API: command not done
Fusion036: Link broken!
Program received signal SIGSEGV, Segmentation fault.
#0  /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f743b234520]
#1  /lib/x86_64-linux-gnu/libc.so.6(+0x19d7fd) [0x7f743b38f7fd]
#2  /snap/freecad/1202/usr/lib/libFreeCADApp.so(+0x4b38e8) [0x7f743d7ec8e8]
#3  0x7f743d7fe12b in App::PropertyString::setValue(char const*) from /snap/freecad/1202/usr/lib/libFreeCADApp.so+0x41b
#4  0x7f73b37d536f in Part::Box::Restore(Base::XMLReader&) from /snap/freecad/1202/usr/lib/Part.so+0x1ff
#5  0x7f743d5cb68d in App::Document::readObjects(Base::XMLReader&) from /snap/freecad/1202/usr/lib/libFreeCADApp.so+0x76d
#6  0x7f743d5cd8fa in App::Document::importObjects(Base::XMLReader&) from /snap/freecad/1202/usr/lib/libFreeCADApp.so+0x16a
#7  0x7f743e0e1e6d in Gui::MergeDocuments::importObjects(std::istream&) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x23d
#8  0x7f743e6499cd in Gui::MainWindow::insertFromMimeData(QMimeData const*) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x1ad
#9  0x7f743e140314 in StdCmdPaste::activated(int) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x84
#10  0x7f743e139f46 in Gui::Command::_invoke(int, bool) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x246
#11  0x7f743e13a406 in Gui::Command::invoke(int, Gui::Command::TriggerSource) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x126
#12  /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5(+0x2f48f4) [0x7f743b9938f4]
#13  0x7f743c4eec26 in QAction::triggered(bool) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x46
#14  0x7f743c4f191c in QAction::activate(QAction::ActionEvent) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5+0xac
#15  0x7f743e1aad1f in Gui::ShortcutManager::onTimer() from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0xdf
#16  0x7f743e1ab8e3 in Gui::ShortcutManager::checkShortcut(QObject*, QKeySequence const&) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x533
#17  0x7f743e1ac7d4 in Gui::ShortcutManager::eventFilter(QObject*, QEvent*) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0xf4
#18  0x7f743b95ba6e in QCoreApplicationPrivate::sendThroughApplicationEventFilters(QObject*, QEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5+0x8e
#19  0x7f743c4f57a0 in QApplicationPrivate::notify_helper(QObject*, QEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5+0xc0
#20  0x7f743e0d69b0 in Gui::GUIApplication::notify(QObject*, QEvent*) from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0xd0
#21  0x7f743b95be4a in QCoreApplication::notifyInternal2(QObject*, QEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5+0x13a
#22  0x7f743bdb5164 in QShortcutMap::dispatchEvent(QKeyEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Gui.so.5+0x454
#23  0x7f743bdb5b03 in QShortcutMap::tryShortcut(QKeyEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Gui.so.5+0x63
#24  0x7f743bd5b2c2 in QWindowSystemInterface::handleShortcutEvent(QWindow*, unsigned long, int, QFlags<Qt::KeyboardModifier>, unsigned int, unsigned int, unsigned int, QString const&, bool, unsigned short) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Gui.so.5+0x192
#25  0x7f743bd8027f in QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Gui.so.5+0x8f
#26  0x7f743bd57dbc in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Gui.so.5+0xac
#27  /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/../../../libQt5XcbQpa.so.5(+0x76f5e) [0x7f74353d3f5e]
#28  /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x26b) [0x7f74393fbd3b]
#29  /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0xab258) [0x7f7439451258]
#30  /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_iteration+0x33) [0x7f74393f93e3]
#31  0x7f743b9b5338 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5+0x68
#32  0x7f743b95a76b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5+0x12b
#33  0x7f743b962d14 in QCoreApplication::exec() from /snap/freecad/1202/kf5/usr/lib/x86_64-linux-gnu/libQt5Core.so.5+0x94
#34  0x7f743e011c88 in Gui::Application::runApplication() from /snap/freecad/1202/usr/lib/libFreeCADGui.so+0x6b8
#35  /snap/freecad/1202/usr/bin/FreeCAD(+0x9791) [0x55d138e78791]
#36  /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f743b21bd90]
#37  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7f743b21be40]
#38  /snap/freecad/1202/usr/bin/FreeCAD(+0x9b55) [0x55d138e78b55]


```

**Comment by luzpaz** [comment]
This is a terrible bug and we need to address it before next minor release.

**Comment by hyarion** [comment]
Does anyone know how to reproduce the issue?

**Comment by axxel** [comment]
Yes I do. See my original problem description. I have not checked whether current HEAD behaves any differently with regards to this. Please note that this issue is not about fixing the SIGSEGV (I believe the particular one I ran into is fixed) but about a flaw in the recovery process in case a crash happened during saving leading to a truncated project file.

My suggestion for a fix would be to simply write every saved file as a temporary first (say `*.FCTmp`) and only after this is completed, then do a `FCStd` -> `FCBak` swap followed by `FCTmp` -> `FCStd` swap. If there is a `FCTmp` found on open-file it is almost certainly corrupt and might simply be discarded or maybe ask the user to send the file in for debugging the crash that let it sit there.

**Comment by hyarion** [comment]
@axxel in the original problem description you wrote "do some work". I've done some work but I haven't had this issue. I can only assume that we're doing some work differently.
It is really difficult to debug when there's no deterministic way of triggering the bug with an example file. However, due to the nature of the bug, I do understand that it's really difficult to produce example + steps.

As an alternative, would it be possible to create a video recording of when you're working and share what leads up to the crash? Maybe that could help finding what causes it?

**Comment by axxel** [comment]
> would it be possible to create a video recording of when you're working and share what leads up to the crash?

There still seems to be a misunderstanding: This is not about the crash itself. This is about the data loss during restart/reopening a broken file _after_ the crash happened. Which particular crash is irrelevant. This has (almost certainly) nothing to do with the OS. There are clear steps above on how to "reproduce" the issue without any crash required.

**Comment by hyarion** [comment]
@axxel Yes, I understand that it is about dataloss and not about a specific crash. The problem is that I've not been able to reproduce a dataloss at sigsegv. This means that I cannot debug the dataloss.

If you, or someone else, can find an easy way to get to a dataloss situation, then we can do the same thing under a test environment and see why the dataloss happens and fix that dataloss issue.

**Comment by axxel** [comment]
@hyarion have you tried [those steps](https://github.com/FreeCAD/FreeCAD/issues/18044#issuecomment-2496822295)?

**Comment by hyarion** [comment]
> @hyarion have you tried [those steps](https://github.com/FreeCAD/FreeCAD/issues/18044#issuecomment-2496822295)?

Thanks for pointing to that. I made some bad assumptions the first time I read it.

So there's three issues (ordered chronological):
1. Crash of application
2. Crash causes data loss
3. If file is corrupted, recovery files won't be loaded as corrupted file will be newer

From what I've seen, 1 and 2 seems to be that it crashes when writing the a name in an xml field to the file (file ends with something like `name="` without closing `"`. Finding why this happens would be really good. But as noted earlier a crash should not end up with a broken file. User files are the most sacred for a program.

Then we have point 3. As a separate, but just as important issue. If a file is corrupt, then surely the recovery file should be possible to load, even if the original file has been changed (corrupted) afterwards.

I haven't verified anything of the above so first step is to do that, then I'll look into number three here.

(Note I'm ok if someone else wants to fix this issue)

**Comment by axxel** [comment]
Looks like now we are on the same page. Point 3 is the only thing I've been talking about in this issue report as this makes all the difference between loosing a couple minutes of work (since last auto-save) and loosing all your work.

**Comment by Lorthiz** [comment]
Just to add to the pile. Worked on a project for a few days. Today I started working in the Assembler and it corrupted my file. I am really happy to be a beta tester for a feature included in 1.0 and not being informed it's not finished and may be even broken. Way to go!

Simple workflow that broke my file:
1) Added 2 new parts to the assembler;
2) Tried positioning one using fixed joint;
3) Noticed it's wrong orientation;
4) Ctrl - Z or Esc, or something;
5) Part disappears;
6) Save;
7) Relad;
8) File broken;

![Image](https://github.com/user-attachments/assets/42a0d8c7-bd3e-47c6-b045-6cdc2598ad3c)

**Comment by luzpaz** [comment]
Seeing a lot of complaining about this bug on different social media posts. This bug is super problematic and should be considered high priority, IMHO.

**Comment by FlachyJoe** [comment]
The logic could be:

1. Check for recovery files as now
2. When opening a document fails then ask the recovery system for a corresponding save

The recovery logic should be separated from the dialog code.

**Comment by luzpaz** [comment]
soft bump

**Comment by FlachyJoe** [comment]
Wouldn't this be an opportunity to merge the backup and the recovery systems in a "time machine"? This new feature could address #6697 and #10032 too.

[What about a "Time Machine" which manages backup and recovery files? #21104](https://github.com/FreeCAD/FreeCAD/discussions/21104)

**Comment by chennes** [comment]
As @FlachyJoe notes, the long-term solution to this is to break apart the `DocumentRecovery` classes to separate the GUI part from the business logic, allowing later re-use of the recovery code, which right now is intertwined with the dialog (and is completely undocumented).

I poked at it a bit today and for this particular sort of corruption there is a reasonably simple "fix": right now, the `isValidProject()` method is only checking to see if `App::ProjectFile::loadDocument()` succeeds. This basically makes sure that `Document.xml` is OK, but ignores `GuiDocument.xml` since that file is effectively optional. However, _if it exists_, it should be valid XML, and that is **not** being checked. So we could modify `isValidProject()` to also check that. This would eliminate the `Ignore recovery file` message seen in these particular cases, presenting the user with a recovery option when their file is corrupt. **HOWEVER** it would still be up to the user to know they needed to *use* that recovery option, because right now the recovery GUI does *not* tell users that their file is corrupted, it just presents the recovery dialog. So:

1. Quick fix: modify `isValidProject()` to check GuiDocument.xml too
2. Medium fix: also make the recovery window tell users when the primary document is corrupted
3. Long term: refactor these classes into GUI and non-GUI work, and modify file loading itself to check for recovery files if it fails.

**Comment by xtemp09** [comment]
I saved file, corrupted it and _FreeCAD_ does not offer me anything when opening the corrupt file. Is this behavior normal? Should the issue be reopened?


```ruby
OS: Ubuntu 24.04.3 LTS (KDE/plasma/xcb)
Architecture: x86_64
Version: 26.3.0dev.48355 (Git)
Build date: 2026/08/31 04:23:34
Build type: Debug
Branch: master
Hash: 433b2b142dc89b12d20dacb92217f7378f4bbfc9
Python 3.12.3, Qt 6.11.0, Coin 4.0.10 (bundled), Pivy 0.6.11 (bundled), Vtk 9.1.0, boost 1_83, Eigen3 3.4.0, PySide 
shiboken , SMESH 7.7.1.0, xerces-c 3.2.4, Clipper2 , OCC 7.6.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
QPA/File dialog/Color dialog: xcb/via Qt/via Qt
Navigation Style/Orbit Style/Rotation Mode: Blender/Rounded Arcball/Window center
Logical DPI/Physical DPI/Pixel Ratio: 96/96.0584/1
OpenGL version/vendor/renderer: 4.3 (Compatibility Profile) Mesa 25.2.8-0ubuntu0.24.04.2/VMware, Inc./SVGA3D; build: RELEASE;  LLVM;
```
